src/material.cpp

```diff
@@ -469,13 +485,163 @@ void pyne::Material::write_hdf5(std::string filename, std::string datapath,
   H5Tclose(attrtype);
 
+  delete[] mat_data;
+
+}
+
+
+void pyne::Material::write_hdf5(std::string filename, std::string datapath,
+                                float row, int chunksize) {
+  hid_t material_grp_id;  // Holder of HDF5 Id of the "/material" group
+  hid_t data_id;  // Holder of HDF5 Id of the data group to write the data
+                  // (located in "/material/datapath")
+
+  // Turn off annoying HDF5 errors
+  H5Eset_auto2(H5E_DEFAULT, NULL, NULL);
+
+  // Set file access properties so it closes cleanly
+  hid_t fapl;
+  fapl = H5Pcreate(H5P_FILE_ACCESS);
+  H5Pset_fclose_degree(fapl, H5F_CLOSE_STRONG);
+  // Create new/open datafile.
+
+  // This complicated algorythm is required to allow backward compatibility with
+  // previous version of write_hdf5 (where data were written in a hdf5 DATASET)
+  // FILE EXIST ?
+  //    NO -> create file with a "/material" group, write the data in it
+  //    YES:
+  //      - DATAPTH exist:
+  //        - YES && != /material -> Detect NUCPATH + old write_hdf5
+  //        - NO -> CONTINUE
+  //          - "/material" EXIST:
+  //            - NO -> create "/material" group and write everyting in it
+  //            - YES:
+  //              - "/material" is a DATASET -> detect NUCPATH + old write_hdf5
+  //              - "/material" is a GROUP -> add data in it
+  //              - "/material" isn't a GROUP nor a DATASET -> fail and complain
+
+  hid_t db;
+  if (!pyne::file_exists(filename)) {
+    // Create the file
+    db = H5Fcreate(filename.c_str(), H5F_ACC_TRUNC, H5P_DEFAULT, fapl);
+
+  } else {
+    bool ish5 = H5Fis_hdf5(filename.c_str());
+    if (!ish5) throw h5wrap::FileNotHDF5(filename);
+    db = H5Fopen(filename.c_str(), H5F_ACC_RDWR, fapl);
+
+    // if datapath != /material and exist -> old format
+    if (h5wrap::path_exists(db, datapath) && datapath != "/material") {
+      std::string nucpath;
+      hid_t data_set = H5Dopen2(db, datapath.c_str(), H5P_DEFAULT);
+
+      // Grab the nucpath
+      if (detect_nuclidelist(data_set, nucpath)) {
+        H5Fclose(db);
+        deprecated_write_hdf5(filename, datapath, nucpath, row, chunksize);
+        return;
+      } else {  // can't find a valid nuclide list path form datapath... fail
+        throw std::runtime_error(
+            "Can't find the nuclide list path in the existing datapath. Can't "
+            "add your material to the datapath.");
+      }
+    }
+  }
+  
+  // Check if /material exist and what type it is
+  herr_t status;
+  H5O_info_t object_info;
+  status = H5Eset_auto2(H5E_DEFAULT, NULL, NULL);
+  status = H5Oget_info_by_name(db, "/material", &object_info, H5P_DEFAULT);
+  // check if "/material" exists
+  if (status == 0) {
+    if (object_info.type ==
+        H5O_TYPE_DATASET) {  // "/material" is a dataset -> old format use
+                             // deprecated write_hdf5
+      H5Fclose(db);
+      deprecated_write_hdf5(filename, datapath, "/nucid", row, chunksize);
```

Comment:
Is it worth it to overload `deprecated_write_hdf5` to have a version that takes a `db` and one that takes a `filename`?

It looks awkward to close this `db` and then open it again in the method

Reply:
as I rename it, I can probably change it to directly take a `db` instead of a filename

Reply:
You might need/want a version that takes a filename as well - at least looking at the use cases above